prow/spyglass/gcsartifact_fetcher_test.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package spyglass
+
+import (
+	"testing"
+)
+
+func TestNewGCSJobSource(t *testing.T) {
+	testCases := []struct {
+		name        string
+		src         string
+		exJobPrefix string
+		exBucket    string
+		exName      string
+		exBuildID   string
+		expectedErr error
+	}{
+		{
+			name:        "Test standard GCS link",
+			src:         "gs://test-bucket/logs/example-ci-run/403",
+			exBucket:    "test-bucket",
+			exJobPrefix: "logs/example-ci-rin/403",
+			exName:      "example-ci-run",
+			exBuildID:   "403",
+			expectedErr: nil,
+		},
+		{
+			name:        "Test GCS link with trailing /",
+			src:         "gs://test-bucket/logs/example-ci-run/403/",
+			exBucket:    "test-bucket",
+			exJobPrefix: "logs/example-ci-rin/403",
+			exName:      "example-ci-run",
+			exBuildID:   "403",
+			expectedErr: nil,
+		},
+		{
+			name:        "Test GCS link with org name",
+			src:         "gs://test-bucket/logs/sig-flexing/example-ci-run/403",
+			exBucket:    "test-bucket",
+			exJobPrefix: "logs/example-ci-rin/403",
+			exName:      "example-ci-run",
+			exBuildID:   "403",
+			expectedErr: nil,
+		},
+		{
+			name:        "Test invalid GCS link",
+			src:         "garbage/test-bucket/logs/sig-flexing/example-ci-run/403",
+			expectedErr: ErrCannotParseSource,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			jobSource, err := newGCSJobSource(tc.src)
+			if err != nil && err != tc.expectedErr {
+				t.Errorf("Expected err: %v, got err: %v", err, tc.expectedErr)
+			}
+			if tc.exBucket != jobSource.bucket {
+				t.Errorf("Expected bucket %s, got %s", tc.exBucket, jobSource.bucket)
+			}
+			if tc.exName != jobSource.jobName {
+				t.Errorf("Expected name %s, got %s", tc.exName, jobSource.jobName)
+			}
+			if tc.exName != jobSource.jobName {
+				t.Errorf("Expected name %s, got %s", tc.exName, jobSource.jobName)
+			}
+		})
+	}
+}
+
+// Tests listing objects associated with the current job in GCS
+func TestArtifacts_ListGCS(t *testing.T) {
```

Comment:
@fejta made the comment as well, but whenever possible I want to be able to scope the amount my brain needs to think to everything between the `{}` on this test function only. The more I need to read outside, the harder it is for me to consider what the implications are and the harder it is to enforce some invariant. Instead of enforcing "we list artifacts correctly" we are now enforcing "given the correct global setup for our fake, given that testmain still sets the global, given that .... " and that makes it easier to break this test without actually changing the functionality it is supposed to be testing. Can you fabricate simple `jobSource`s in this test that you construct to return a simple list, etc?